### ISSUE TITLE
HTML mode per block

### ISSUE THREAD

**Issue by mtias** [WordPress/gutenberg#2794]
Instead of offering the full text mode so prominently as the way to edit HTML, what if we add a mechanism to every block (in the advanced settings of the inspector) to "edit as HTML"? I'd imagine it would convert the block to an HTML block, which supports preview. This also solves the issue of switching and not knowing which HTML applies to what.

The text mode would still exist for different reasons, but perhaps we can make it way less prominent.

cc @jasmussen @karmatosed @folletto

**Comment by jasmussen** [comment]
I kind of love this idea. It seems like it would make any "document surgery" way more convenient, while still actually behaving like a traditional text mode for any old posts being edited in the Classic Text block. We could potentially put the full-on text mode inside the ellipsis menu mocked up in #2460

**Comment by mtias** [comment]
> We could potentially put the full-on text mode inside the ellipsis menu mocked up in 

Indeed.

**Comment by folletto** [comment]
I think it's worth an exploration! :) 

And I agree making it less prominent makes sense. In that case I'd consider a keyboard shortcut to switch between the two, because, y'know, power users :)

**Comment by karmatosed** [comment]
I like this idea. Would it be the case that only that block would get messed up if they edited wrongly? My thinking there is that avoids them wrecking by accident the entire document. Some way to just delete per block could be nice.

It also avoids the 'sea' of text that text mode presents. Being able to just go into the one you want as a huge benefit. I would +1 to a keyboard shortcut but also have a non keyboard option, just not that obvious. Obviousness is where this option creates a trap.

**Comment by folletto** [comment]
Yep, sorry if I wasn't clear: the keyboard shortcut should be on top of existing UI as above, not replacing it as the unique way to access it.

**Comment by mtias** [comment]
> Would it be the case that only that block would get messed up if they edited wrongly? My thinking there is that avoids them wrecking by accident the entire document.

Yes, exactly. And it isolates the HTML comments too, so you don't see them.

**Comment by jwold** [comment]
Wanting to clarify if my understanding of what this ticket is trying to accomplish is correct. Added some annotations to the design to validate. 

![annotation](https://user-images.githubusercontent.com/3411173/30928838-892d44ac-a371-11e7-8cfa-1fb630567daa.png)

**Comment by mtias** [comment]
@jwold that's correct. Though the "Edit as HTML" might make more sense below the trash icon on the block instead of in the inspector. Something to play with.

**Comment by jwold** [comment]
> @jwold that's correct. Though the "Edit as HTML" might make more sense below the trash icon on the block instead of in the inspector. Something to play with. 

That could work. I'm guessing this icon could make sense:

<img width="39" alt="screen shot 2017-09-27 at 11 14 34 am" src="https://user-images.githubusercontent.com/3411173/30929976-113536a4-a375-11e7-8894-8e91e30477da.png">

**Comment by mtias** [comment]
There's a special HTML code icon that we use for the HTML block.

**Comment by jasmussen** [comment]
I mistakenly posted mockups on the PR for this (https://github.com/WordPress/gutenberg/pull/2797#issuecomment-333073887) as opposed to here. But @iseulde raises a great concern on that post too (https://github.com/WordPress/gutenberg/pull/2797#issuecomment-333206704) so this is a good opportunity to refresh the mockups a bit. CC: @jwold 

Ella's comment:

> @jasmussen Honestly my first thought with the ellipsis (both at block level and document level) was... not another hidden area. We already have the inspector, which kind of acts as a place were all the advanced stuff goes. Also for the one at the top next to publish, is that not staff that could go in the document inspector? The generic cog and ellipsis next to each other feel quite strange. What doe those even mean? What's the difference? I understand the need for something at the inline block level, where we need a solution for overflow, I'm still wondering though if we can't just put an ellipsis there that would open up the inspector (and get rid of the other one).

Solid points. 

The inspector is where "advanced stuff" goes, so why do we need an ellipsis on the quick toolbar? 

My answer to this is to suggest that items in the ellipsis aren't necessarily advanced, but rather _overflow_ — items from the quick toolbar that didn't fit. [These](https://user-images.githubusercontent.com/1204802/31008745-73bba2d6-a505-11e7-9e1b-b7ba8579c19a.png) [two](https://user-images.githubusercontent.com/1204802/31008749-75820650-a505-11e7-921a-6b9a91070199.png) mockups did not help convey that message, pretend they don't exist. 

Here are better mockups. This is the desktop, there's room for all items in the quick toolbar:

![desktop there s room](https://user-images.githubusercontent.com/1204802/31067423-2d92e418-a753-11e7-9c41-a734fa28ee80.png)

On mobile, the items pop off the right as the screen gets smaller. First, the HTML editor mode and footnote (example button) pop off, then if the screen grows smaller, the strikethrough also:

<img width="417" alt="screen shot 2017-10-02 at 09 21 47" src="https://user-images.githubusercontent.com/1204802/31067459-4befaf18-a753-11e7-8160-e273aa969728.png">

Does that make sense?

**Comment by iseulde** [comment]
@jasmussen Yes, as noted in my comment, I understand that this is for overflow, whereas the other is for advanced items, but I still wonder if we can't blur that boundary and overflow into the "inspector", whatever the form is (mobile/desktop)? I also'd like to rephrase that the issue with the block toolbar/inspector is relevant for the document toolbar/inspector as well. There we even have a cog and an ellipsis right next to each other, which I personally find confusing. When I need something, and I don't know where it is, I'll need to look under two icons which just both tell me "here is general stuff".

**Comment by jasmussen** [comment]
It's a solid argument. I'll keep thinking about what the best approach is, here. As it is right now, it's very much emulating mobile patterns.

**Comment by afercia** [comment]
Just as a reminder: a "text mode" with a simple textarea is a must-have for accessibility.

**Comment by mtias** [comment]
Done in #2797. Nice work @youknowriad !
